MIMIC

nirhod opened the pull request.

nirhod committed:
MIMIC

AlonSh commented on mimic.py:
```diff
@@ -0,0 +1,153 @@
+class Node(object):
+
```
No need for a blank line after the class declaration, but *when* you'll add class doc string it will be OK.

AlonSh commented on mimic.py:
```diff
@@ -0,0 +1,153 @@
+class Node(object):
```
No need to inherit from object in python3

AlonSh commented on mimic.py:
```diff
@@ -0,0 +1,153 @@
+        return extended
+
+    def _get_node(self, key: int) -> Node:
+        current = self.first_node
```
What happens if key > length?
Also in other methods

AlonSh commented on mimic.py:
```diff
@@ -0,0 +1,153 @@
+
+    def pop(self, index: int):
+        if index == 0:
+            poped_node = self.first_node
```
Typo `poped`

AlonSh reviewed (changes requested):
Looks really good!

nirhod committed:
MIMIC fix 1
Most of the functions are the same as list functions so I didn't add docstring.
If I should add docstring, please tell me

ofir123 commented on mimic.py:
```diff
@@ -0,0 +1,179 @@
+
+    def _get_node(self, key: int) -> Node:
+        """
+        Works like MyList[key] but return the node instead of the value
```
It's weird to reference another function in docstring, when that function actually calls this one.

ofir123 commented on mimic.py:
```diff
@@ -0,0 +1,179 @@
+            raise IndexError('list index is out of range')
+        current = self.first_node
+        i = 0
+        while current:
```
Why not use `NodeIterator`?

nirhod replied:
Because NodeIterator return the value but we need the node itself.
(Node has 2 attributes: value and next)

ofir123 commented on mimic.py:
```diff
@@ -0,0 +1,179 @@
+            self._get_node(len(self) - 1).next_node = Node(value)
+        self.length += 1
+
+    def count(self, value):
```
Nice!

ofir123 commented on mimic.py:
```diff
@@ -0,0 +1,179 @@
+        self.first_node = new.first_node
+
+    def sort(self):
+        l = sorted(self)
```
This is a nice opportunity for an algorithm!
Can you try sorting with bubble sort?

ofir123 commented on mimic.py:
```diff
@@ -0,0 +1,179 @@
+        self.extend(l)
+
+    def __repr__(self):
+        return str([i for i in self])
```
This won't be printed like Python's string. (same for `__str__`)

nirhod replied:
I don't think I understand.
When I print it, it looks like I print a list.

ofir123 reviewed (changes requested):
Very nice implementation!

nirhod committed:
Mimic: fixed docstring and added bubble sort

ofir123 commented on mimic.py:
```diff
@@ -0,0 +1,207 @@
+            current_node = self.first_node
+
+    def __repr__(self):
+        return str([i for i in self])
```
List comprehension in not allowed (in this exercise)!

ofir123 commented on mimic.py:
```diff
@@ -0,0 +1,207 @@
+            new.append(self[i])
+        self.first_node = new.first_node
+
+    def sort(self):
```
When implementing a specific algorithm, it's important to add a little documentation.

ofir123 commented on mimic.py:
```diff
@@ -0,0 +1,207 @@
+                next_node = current_node.next_node
+                if current_node.data > current_node.next_node.data:
+                    swap_occurred = True
+                    current_node.data, current_node.next_node.data = current_node.next_node.data, current_node.data
```
Try to avoid using double assignments.
It's very hard to read, and not really necessary.

ofir123 commented on mimic.py:
```diff
@@ -0,0 +1,207 @@
+            swap_occurred = False
+            while current_node.next_node:
+                next_node = current_node.next_node
+                if current_node.data > current_node.next_node.data:
```
Why not `current_node.data > next_node.data`?

nirhod committed:
Added documentation to sort
Fixed double assignment
Fixed using list comperhension in __repr__

nirhod committed:
Fixed sort's docstring

AlonSh reviewed (approved):
Still missing some docs.
